Add a parse_semver filter

JensHeinrich opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This filter uses the SemanticVersion class from utils.version

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
filters

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

Commit:
Add a parse_semver filter

This filter uses the SemanticVersion class from utils.version

bmillemathias commented:
Hi @JensHeinrich 

Thanks for your patch, to complete it we need also a changelog file and some tests to validate the code does not break over time, look to https://docs.ansible.com/ansible/latest/community/development_process.html#making-your-pr-merge-worthy to get an idea how to do this.

Commit:
Add a changelog fragment

JensHeinrich commented:
@bmillemathias on it

bmillemathias commented:
> @bmillemathias on it

Hi, the unit / integration test is missing (test the `semver` filter works but also testing failure behavior)
also could you look to the errors reported by the CI tests.

Best

Commit:
Add error handling

Commit:
Add initial tests for parse_semver filter

Commit:
Add type check

Commit:
Fix tests

JensHeinrich commented:
@bmillemathias Already on it
Currently cleaning the formatting tests

Commit:
Fix pep8 formatting

Commit:
Replace dangerous default value

Fix is described here https://satran.in/b/python--dangerous-default-value-as-argument

Commit:
Remove trailing new line

Commit:
Add space around operator

Commit:
Fix pep8

Commit:
Fix pep8 missing blank lines

Commit:
Change return

mattclay commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -342,6 +343,37 @@ def combine(*terms, **kwargs):
+
+def parse_semver(
+        vstring,
+        keys=None
```
What is the reason for adding keys here? The keys needed could be chosen after using the filter.

JensHeinrich replied:
My idea was to support directly asking for a part like the url filter enables

mattclay reviewed (commented):
It would be good to add this to the docs.

bcoca commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -342,6 +343,37 @@ def combine(*terms, **kwargs):
 
+SEMVER_VALID_KEYS = ['major', 'minor', 'patch', 'prerelease', 'buildmetadata']
+# This list can be generated from the SEMVER_RE expression by
+# `SEMVER_VALID_KEYS=list(SEMVER_RE.groupindex.keys())` but is expected to stay static
```
generate it as frozenset at class level, to avoid hardcoded and redefining per invocation

JensHeinrich replied:
as in the version class?

bcoca replied:
yes

Commit:
Move semver keys to frozenset

Commit:
Add documentation

Commit:
Add explicit version examples

Commit:
Move return of the filter to comments

Commit:
Add python2.6 compatible syntax

Source: https://stackoverflow.com/questions/4514162/collecting-key-value-pairs-in-python-2-6

JensHeinrich commented:
Did a rebase on the current devel branch

bcoca reviewed (changes requested):
also:
 - as mentioned by matt, no need to do the filtering
 -  docs should now live in it's own yaml file https://github.com/ansible/ansible/pull/79056
 -  version_added should be the 'core' version, 2.15 at this point

Akasurde commented:
@JensHeinrich Are you still working on this? 

needs_info

JensHeinrich commented:
I'

> @JensHeinrich Are you still working on this?
> 
> needs_info

I'm trying to get back into it.

Commit:
Add author

Commit:
Remove filtering.

JensHeinrich commented:
> * as mentioned by matt, no need to do the filtering
done
> * docs should now live in it's own yaml file [document adjacent docs (sidecar) #79056](https://github.com/ansible/ansible/pull/79056)
done
> * version_added should be the 'core' version, 2.15 at this point
done

Commit:
Remove keys parameter

nitzmahone commented:
Hi!

Thank you very much for your submission to Ansible. It means a lot to us that you've taken the time to contribute.

Since Ansible 2.10 we are no longer accepting new modules/plugins into Ansible core.  However, we recommend looking into providing this functionality through Ansible Galaxy via Ansible Collections. You can find more information about collections at:

* https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html.

Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
The mailing list and irc are great ways to ask questions, or post if you don't think this particular issue is resolved.

See  this page for a complete and up to date list of communication channels and their purposes:

* https://docs.ansible.com/ansible/latest/community/communication.html

Thank you once again for this and your interest in Ansible!
